src/WebJobs.Extensions.DurableTask/AzureStorageDurabilityProviderFactory.cs

```diff
@@ -1,150 +1,167 @@
-
-            var settings = new AzureStorageOrchestrationServiceSettings
-            {
-                StorageConnectionString = resolvedStorageConnectionString,
-                TaskHubName = taskHubNameOverride ?? this.options.HubName,
-                PartitionCount = this.azureStorageOptions.PartitionCount,
-                ControlQueueBatchSize = this.azureStorageOptions.ControlQueueBatchSize,
-                ControlQueueBufferThreshold = this.azureStorageOptions.ControlQueueBufferThreshold,
-                ControlQueueVisibilityTimeout = this.azureStorageOptions.ControlQueueVisibilityTimeout,
-                WorkItemQueueVisibilityTimeout = this.azureStorageOptions.WorkItemQueueVisibilityTimeout,
-                MaxConcurrentTaskOrchestrationWorkItems = this.options.MaxConcurrentOrchestratorFunctions,
-                MaxConcurrentTaskActivityWorkItems = this.options.MaxConcurrentActivityFunctions,
-                ExtendedSessionsEnabled = this.options.ExtendedSessionsEnabled,
-                ExtendedSessionIdleTimeout = extendedSessionTimeout,
-                MaxQueuePollingInterval = this.azureStorageOptions.MaxQueuePollingInterval,
-                TrackingStoreStorageAccountDetails = GetStorageAccountDetailsOrNull(
-                    this.connectionStringResolver,
-                    this.azureStorageOptions.TrackingStoreConnectionStringName),
-                FetchLargeMessageDataEnabled = this.azureStorageOptions.FetchLargeMessagesAutomatically,
-                ThrowExceptionOnInvalidDedupeStatus = true,
-            };
-
-            if (!string.IsNullOrEmpty(this.azureStorageOptions.TrackingStoreNamePrefix))
-            {
-                settings.TrackingStoreNamePrefix = this.azureStorageOptions.TrackingStoreNamePrefix;
-            }
-
-            return settings;
-        }
-
-        private static StorageAccountDetails GetStorageAccountDetailsOrNull(IConnectionStringResolver connectionStringResolver, string connectionName)
-        {
-            if (string.IsNullOrEmpty(connectionName))
-            {
-                return null;
-            }
-
-            string resolvedStorageConnectionString = connectionStringResolver.Resolve(connectionName);
-            if (string.IsNullOrEmpty(resolvedStorageConnectionString))
-            {
-                throw new InvalidOperationException($"Unable to resolve the Azure Storage connection named '{connectionName}'.");
-            }
-
-            return new StorageAccountDetails
-            {
-                ConnectionString = resolvedStorageConnectionString,
-            };
-        }
-    }
- }
+        }
+
+        // This method should not be called before the app settings are resolved into the options.
+        // Because of this, we wait to validate the options until righ before building a durability provider, rather
+        // than in the Factory constructor.
+        private void EnsureOptionsAreValid()

```

Comment:
I almost wonder if you should name this `Initialize()` or `EnsureIsInitialized()`. It kinda seems like that's what this method is doing - not just validating.